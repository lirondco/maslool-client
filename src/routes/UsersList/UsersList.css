.UsersList {
  margin: 2%;
  padding: 2%;
}

.UsersList h2 {
  margin-bottom: 1%;
}

hr {
  border: 1px solid;
}

.UsersList label {
    font-size: larger;
}
