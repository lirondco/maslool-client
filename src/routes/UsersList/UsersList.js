import React, { Component } from 'react'
import UserComponent from '../../components/UserComponent/UserComponent'
import ModeratorApiService from '../../services/moderator-api-service'
import { Input, Label } from '../../components/Utils/Utils'
import './UsersList.css'

export default class UsersList extends Component {
    state = {
        users: null,
        error: null,
        searchResult: null,
    }

    getUsers = () => {
        ModeratorApiService.getUsers()
            .then(users => this.setState({ users, searchResult: users }))
            .catch(error => this.setState({ ...error }))
    }

    componentDidMount = () => {
        this.getUsers()
    }

    handleToggleBan = (userId) => {
        ModeratorApiService.toggleBanUser(userId)
            .then(() => window.alert('Action completed'))
            .then(() => this.getUsers())
            .catch(error => this.setState({ ...error }))
    }

    handleChange = ev => {
        const searchString = ev.target.value
        if(this.state.users) {
            this.setState({ error: null })
            let searchResult = this.state.users.filter(user => 
                    user.username.toLowerCase().includes(searchString.toLowerCase())
                )
            this.setState({ searchResult })
            if (searchResult.length === 0) {
                this.setState({ error: 'No user with that username'})
            }
        } else {
            this.setState({ error: 'No users available'})
        }
    }

    render() {
        const {error, searchResult } = this.state
        let sortedResult
        if (searchResult !== null) {
            sortedResult = searchResult.sort((a, b)=> a.username.toLowerCase() > b.username.toLowerCase() ? 1: -1)
        }
        return (
            <section className='UsersList'>
                <h2>Users List</h2>
                <hr />
                <Label htmlFor='search_user_box'>
                    Filter by username:
                </Label>
                <Input onChange={this.handleChange} id='search_user_box' name='searchString' />
                <div role="alert">{error && <p className="error">{error}</p>}</div>
                <ul className='user_list'>
                    {sortedResult && sortedResult.map(user => 
                        <li key={user.id}>
                            <UserComponent user={user} onBanUser={() => this.handleToggleBan(user.id)}/>
                        </li>
                        )}
                </ul>
            </section>
        )
    }

}