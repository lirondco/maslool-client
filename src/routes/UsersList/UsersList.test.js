import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import UsersList from "./UsersList";

describe(`UsersList route component`, () => {
  it("renders the complete page when state is empty", () => {
    const wrapper = shallow(<UsersList />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("renders the complete page when state is not empty", () => {
    const wrapper = shallow(<UsersList />);
    const testsUsers = [
      {
        id: 2,
        email: "p.star@bikinibottom.net",
        username: "patrick",
        join_date: "2021-03-11T06:47:06.477Z",
        admin: false,
        banned: true,
        banned_by: "sandy",
      },
      {
        id: 3,
        email: "sandy.cheeks@squirrelmail.com",
        username: "sandy",
        join_date: "2021-03-11T06:47:06.477Z",
        admin: true,
        banned: false,
        banned_by: null,
      },
      {
        id: 6,
        email: "s.tentacles@krustykrab.com",
        username: "squidward",
        join_date: "2021-03-11T06:47:06.477Z",
        admin: false,
        banned: false,
        banned_by: null,
      },
    ]
    wrapper.setState({
      users: testsUsers,
      error: null,
      searchResult: testsUsers,
    });
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
