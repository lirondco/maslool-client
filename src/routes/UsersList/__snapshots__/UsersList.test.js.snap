// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersList route component renders the complete page when state is empty 1`] = `
<section
  className="UsersList"
>
  <h2>
    Users List
  </h2>
  <hr />
  <Label
    htmlFor="search_user_box"
  >
    Filter by username:
  </Label>
  <ForwardRef
    id="search_user_box"
    name="searchString"
    onChange={[Function]}
  />
  <div
    role="alert"
  />
  <ul
    className="user_list"
  />
</section>
`;

exports[`UsersList route component renders the complete page when state is not empty 1`] = `
<section
  className="UsersList"
>
  <h2>
    Users List
  </h2>
  <hr />
  <Label
    htmlFor="search_user_box"
  >
    Filter by username:
  </Label>
  <ForwardRef
    id="search_user_box"
    name="searchString"
    onChange={[Function]}
  />
  <div
    role="alert"
  />
  <ul
    className="user_list"
  >
    <li
      key="2"
    >
      <UserComponent
        onBanUser={[Function]}
        user={
          Object {
            "admin": false,
            "banned": true,
            "banned_by": "sandy",
            "email": "p.star@bikinibottom.net",
            "id": 2,
            "join_date": "2021-03-11T06:47:06.477Z",
            "username": "patrick",
          }
        }
      />
    </li>
    <li
      key="3"
    >
      <UserComponent
        onBanUser={[Function]}
        user={
          Object {
            "admin": true,
            "banned": false,
            "banned_by": null,
            "email": "sandy.cheeks@squirrelmail.com",
            "id": 3,
            "join_date": "2021-03-11T06:47:06.477Z",
            "username": "sandy",
          }
        }
      />
    </li>
    <li
      key="6"
    >
      <UserComponent
        onBanUser={[Function]}
        user={
          Object {
            "admin": false,
            "banned": false,
            "banned_by": null,
            "email": "s.tentacles@krustykrab.com",
            "id": 6,
            "join_date": "2021-03-11T06:47:06.477Z",
            "username": "squidward",
          }
        }
      />
    </li>
  </ul>
</section>
`;
